Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/ADLSGen2Deploy.java

```diff
@@ -0,0 +1,252 @@
+            throw new IllegalArgumentException("Cluster type doesen't support adls gen2 storage type.");
+        }
+
+        http.setDefaultHeaders(headerBuilder.build());
+    }
+
+
+    @Override
+    public URI getUploadDir(@NotNull String rootPath) {
+        return URI.create(rootPath)
+                .resolve(JobUtils.getFormatPathByDate() + "/");
+    }
+
+    @NotNull
+    public Observable<String> deploy(@NotNull File src, @NotNull URI destURI) {
+        //three steps to upload via adls gen2 rest api
+        // 1.put request to create new dir
+        // 2.put request to create new file(artifact) which is empty
+        // 3.patch request to append data to file
+        // 4.patch request to flush data to file
+
+        try {
+            InputStreamEntity reqEntity = new InputStreamEntity(
+                    new FileInputStream(src),
+                    -1,
+                    ContentType.APPLICATION_OCTET_STREAM);
+            BufferedHttpEntity entity = new BufferedHttpEntity(reqEntity);
+            String len = String.valueOf(entity.getContentLength());
+
+            //remove request / end otherwise invalid url response
+            String destStr = destURI.toString();
+            String dirPath = destStr.endsWith("/") ? destStr.substring(0, destStr.length() - 1) : destStr;
+            String filePath = String.format("%s/%s", dirPath, src.getName());
+
+            HttpPut req = new HttpPut(dirPath);
+            return http.request(req, null
+                    , this.createDirReqParams
+                    , this.headerBuilder.setAuthorization(
+                            this.cred.generateSharedKey(req, this.headerBuilder.getHeaderGroup(), this.createDirReqParams)
+                    ).getHeaderList())
+                    .doOnNext(
+                            resp -> {
+                                if (resp.getStatusLine().getStatusCode() != 201) {
+                                    Exceptions.propagate(new UnknownServiceException("Can not create directory to save artifact using ADLSGen2 storage type"));
+                                }
+                            }
+                    )
+                    .map(ignored -> new HttpPut(filePath))
+                    .flatMap(put -> http.request(put, null
+                            , this.createFileReqParams
+                            , this.headerBuilder.setAuthorization(
+                                    this.cred.generateSharedKey(put, this.headerBuilder.getHeaderGroup(), this.createFileReqParams)
+                            ).getHeaderList()))
+                    .doOnNext(
+                            resp -> {
+                                if (resp.getStatusLine().getStatusCode() != 201) {
+                                    Exceptions.propagate(new UnknownServiceException("Can not create artifact using adlsgen2 storage type"));
+                                }
+                            }
+                    )
```

Comment:
The `map`, `flatMap` and `doOnNext` code can be put into a dedicated function to help understanding.